File: 08-streaming/8.4-structured/src/main/scala/structured/SQL.scala
/*

1. Run the app
    $  $SPARK_HOME/bin/spark-submit  --master local[2]   --driver-class-path logging/  \
       --class structured.SQL     target/scala-2.12/structured-streaming_2.12-1.0.jar 
       
2. supply input
     $    cp   clickstream.json    input/1.json
     $    cp   clickstream.json    input/2.json
 */

package structured


import org.apache.spark.sql.functions.lit
import org.apache.spark.sql.SparkSession


object SQL {
  def main(args: Array[String]) {

    val spark = SparkSession.builder.appName("Structured Streaming").
                getOrCreate()
    import spark.implicits._
      
    // Set loglevel to Error
    spark.sparkContext.setLogLevel("ERROR")

    println("### Spark UI available on port : " + spark.sparkContext.uiWebUrl.get.split(':')(2))


    val sample_data = spark.read.json("clickstream.json")
    println ("clickstream schema:")
    sample_data.printSchema

    val schema = sample_data.schema
    println("ClickStream sample schema is : " + schema)
      
    val click_stream = spark.readStream
                       .schema(schema)
                       .json("input/")
      
    // Print out incoming text
    // the 'withColumn' is for debug, so we can identify query outupt
    val query1 = click_stream
                    .withColumn ("query", lit("query1-clickstream"))
                    .writeStream
                    .outputMode("append")
                    .format("console")
                    .queryName("query1-clickstream")
                    .start()

   click_stream.createOrReplaceTempView("clickstream")
      
   val sql_str = """
                    select domain, count(*) as visits, SUM(cost) as spend 
                    from clickstream
                    group by domain
                    order by visits DESC
                    """

    val by_domain_results = spark.sql(sql_str)
      
    val query2 = by_domain_results
            .withColumn ("query", lit("query2-bydomain"))
            .writeStream
            .outputMode("complete")
            .option("truncate", false)
            .format("console")
            .queryName("query2-bydomain")
            .start()
      
    // wait forever until user terminate manually
    spark.streams.awaitAnyTermination() 
    spark.stop()
  }
}

File: 08-streaming/8.4-structured/python/sql.py
"""

1. Run the app
     $  cd <project root dir>
     $  rm -f input/* ; $SPARK_HOME/bin/spark-submit --master local[2] \
            --driver-class-path ../logging/ \
            python/sql.py
            
2. supply input
     $    cp   clickstream.json    input/1.json
     $    cp   clickstream.json    input/2.json
 
"""

from pyspark.sql import SparkSession
from pyspark.sql.functions import lit

# Initialize Spark
spark = SparkSession \
    .builder \
    .appName("StructuredStreamingExample") \
    .getOrCreate()

# Set loglevel to Error
spark.sparkContext.setLogLevel("ERROR")
print('### Spark UI available on port : ' + spark.sparkContext.uiWebUrl.split(':')[2])


sample_data = spark.read.json("clickstream.json")
print ("clickstream schema:")
sample_data.printSchema()

schema = sample_data.schema
print("ClickStream sample schema is : " , schema)



click_stream = spark.readStream  \
               .schema(schema)  \
               .json("input/")


## print out all incoming data
query1 = click_stream \
    .withColumn ("query", lit("query1-clickstream")) \
    .writeStream \
    .outputMode("append") \
    .option("truncate", False) \
    .format("console") \
    .queryName("query1-clickstream") \
    .start()
# ----- end-TODO-2


# register a temp table
click_stream.createOrReplaceTempView("clickstream")


sql_str = (
'select domain, count(*) as visits, SUM(cost) as spend '
'from clickstream '
'group by domain '
'order by visits DESC '
)

# sql_str = """
# select domain, count(*) as visits, SUM(cost) as spend 
# from clickstream
# group by domain
# order by visits DESC
# """

# TODO-3 :
by_domain_results = spark.sql(sql_str)

query2 = by_domain_results \
    .withColumn ("query", lit("query2-bydomain")) \
    .writeStream \
    .outputMode("complete") \
    .option("truncate", False) \
    .format("console") \
    .queryName("query2-bydomain") \
    .start()
# ----- end-TODO-3


spark.streams.awaitAnyTermination() 
spark.stop()
